### ISSUE TITLE
FORMAT function fails on tabulation.

### ISSUE THREAD

**Issue by 2403772980ygy** [clasp-developers/clasp#1348]
**Describe the bug**
When using the `format` function on a list, the tabulation breaks.

**Expected behavior**
`(defun print-list(ls) (format t "~{~a:~10t~a~%~}" ls))`
When using this to output a list like`(list :a 1 :b 2 :c 3)`
It should give me the following, which is the output of clisp:
```
A:        1
B:        2
C:        3
NIL
```

**Actual behavior**
It gives the following output, no matter with `rlwrap` or not:
```
A: 1
B:        2
C:        3
NIL
```
**Code at issue**

```lisp
(defun print-list(ls) (format t "~{~a:~10t~a~%~}" ls))
(print-list (list :a 1 :b 2 :c 3))
```

* Versions or commit hashes of Clasp on which you observed the problem (if you have built from up to date source, saying so is fine, you don't need to dig around for hashes)
commit 6b5234a398e2ae53302300deb9fd8c90ea9321dd built from source on Gentoo
* If relevant, your operating system and other aspects of your computing environment
Gentoo 17.1, llvm and clang using version 13.0.1
* If you believe the compiler may be at fault, relevant declarations, including optimize settings (use `(clasp-cltl2:declaration-information 'optimize)`)
* Any other context about the problem

**Comment by Bike** [comment]
confirmed on aba76e80. problem goes away if the format string starts with a `~&`. odd.

**Comment by 2403772980ygy** [comment]
@Bike Well, I kinda see the issue, other implementations does have the same issue when the two lines of code is copied. I guess that's because when we copy the code we get an line feed as the end of a line. However when using the REPL and type in the code one line after another, we puts an carrier return which then causes the REPL to prompt, which then causes some implementations to break.

You may try typing in the code one line after another, the problem persists, which does not happen on other implementations.

**Comment by Bike** [comment]
i'm sorry, i am not sure i totally understand what you mean by "typing in the code one line after another". but i'm not seeing the problem go away on SBCL with multiple prompts:
```
* (defun print-list(ls) (format t "~{~a:~10t~a~%~}" ls)) 
PRINT-LIST
* (loop (write-string "COMMON-LISP-USER> ") (finish-output) (print (eval (read))))
COMMON-LISP-USER> (print-list '(:a 2 :b  7))
A: 2
B:        7

NIL COMMON-LISP-USER> (print-list '(:a 2 :b 7))
A: 2
B:        7
```

**Comment by 2403772980ygy** [comment]
Hmm, this is interesting.
For me, this happens on sbcl(with and without rlwrap), colzurecl(no rlwrap) and ecl(no rlwrap) but not on clisp(has rlwrap on by default).
Only clisp survived the test. This might not be a bug then.

**Comment by 2403772980ygy** [comment]
I guess that implementing another "virtual" stream for output aside from the

**Comment by 2403772980ygy** [comment]
I guess that implementing another "virtual" stream for output aside from the REPL may be an way out.

**Comment by Bike** [comment]
yeah, if you output to a string stream or something there will be no problem.

**Comment by Bike** [comment]
> Some implementations such as sbcl has this issue probably because the `write-string` use the same "stream" as the `format`. And the bug appears.

Maybe I did not explain clearly enough. That is not the issue. The issue is specific to I/O streams, i.e. streams used both for input and output. Basically, the system (either Clasp, SBCL, or clisp) tracks the column of text _output_ but not the column _input_. So Clasp prints `"COMMON-LISP-USER> "` and notes the column as being the length of that string, 18. Then the user types whatever and hits enter; but this typing and the newline do _not_ influence the column Lisp knows about. So when format comes around, it sees `~10t` and tries to move to at least the 10th column, but it thinks the column is already 18, so it just puts in a space.

In other words the issue is because the stream is used for both input and output but the system only tracks the output column. On an output only stream, as to a file or a string stream, there will be no issue, because there is no input to influence the column.

Moreover, for a stream representing a terminal, it may be difficult or impossible for us to track the column fully because other programs can output to the terminal, for example the shell, and we would have to have a whole terminal interface specifically. I am actually not sure if this is doable without taking over the terminal entirely via ncurses or the like.

I have been able to reproduce the problem on clisp the same as I did on SBCL.

**Comment by yitzchak** [comment]
This is probably why PRINT/PPRINT precede output with a newline, versus a trailing newline.

**Comment by yitzchak** [comment]
@2403772980ygy we have a potential fix.

**Comment by yitzchak** [comment]
Looks like it works.

```
COMMON-LISP-USER> (defun print-list(ls) (format t "~{~a:~10t~a~%~}" ls))
PRINT-LIST
COMMON-LISP-USER> (print-list '(:a 1 :b 2 :c 3))
A:        1
B:        2
C:        3
NIL
```
